lib/src/license_detection/license.dart

```diff
@@ -0,0 +1,247 @@
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:meta/meta.dart';
+import 'package:pana/src/third_party/diff_match_patch/diff.dart';
+import 'package:pana/src/license_detection/token_matcher.dart';
+import 'package:pana/src/license_detection/tokenizer.dart';
+
+import 'crc32.dart';
+
+@sealed
+class License {
+  /// SPDX identifier of the license, is empty in case of unknown license.
+  final String identifier;
+
+  /// Original text from the license file.
+  final String content;
+
+  /// Normalized [Token]s created from the original text.
+  final List<Token> tokens;
+
+  /// Map from [Token.value] to their number of occurrences in this license.
+  final Map<String, int> occurrences;
+
+  License._(this.content, this.tokens, this.occurrences, this.identifier);
+
+  factory License.parse(String identifier, String content) {
+    final tokens = tokenize(content);
+    final table = generateFrequencyTable(tokens);
+    return License._(content, tokens, table, identifier);
+  }
+}
+
+String tokensNormalizedValue(Iterable<Token> tokens) {
+  var buffer = StringBuffer();
+
+  for (var token in tokens) {
+    buffer.write(token.value);
+    buffer.write(' ');
+  }
+
+  return buffer.toString().trimRight();
+}
+
+@sealed
+class NGram {
+  /// Text for which the hash value was generated.
+  final String text;
+
+  /// [CRC-32][1] checksum value generated for text.
+  ///
+  /// [1]: https://en.wikipedia.org/wiki/Cyclic_redundancy_check
+  final int checksum;
+
+  /// Index of the first token in the checksum.
+  final int start;
+
+  /// Index of the last token in the checksum.
+  final int end;
+
+  int get granularity => text.split(' ').length;
+
+  NGram(this.text, this.checksum, this.start, this.end);
+}
+
+/// A [License] instance with generated [nGrams].
+// TODO: Change class name to something more meaningful.
+@sealed
+class PossibleLicense {
+  final License license;
+
+  final List<NGram> nGrams;
+
+  final Map<int, List<NGram>> checksumMap;
+
+  PossibleLicense._(this.license, this.nGrams, this.checksumMap);
+
+  factory PossibleLicense.parse(License license, int granularity) {
+    final nGrams = generateChecksums(license.tokens, granularity);
+    final table = generateChecksumMap(nGrams);
+    return PossibleLicense._(license, nGrams, table);
+  }
+}
+
+/// Contains details regarding the results of corpus license match with unknown text.
+@sealed
+class LicenseMatch {
+  /// Sequence of tokens from input text that were considered a match for the detected [License].
+  final List<Token> tokens;
+
+  /// [Diff]s calculated between target tokens and [license] tokens.
+  final List<Diff> diffs;
+
+  /// Range of [diffs] which represents the diffs between known license and unknown license.
+  ///
+  /// Diffs lying outside of [diffRange] represent the text in unknown license
+  /// that is not a part of the [license](source) text.
+  final Range diffRange;
```

Comment:
Why don't we just make the `diffs` list smaller? What is the point in retaining the things inside of that which are not important?

Also couldn't we just use `diffRange` to reduce the `tokens` list.. by removing some of the first and last tokens?
I'm curious if that isn't what we're doing in the `diffRange` calculation.

Reply:
No, we probably can't reduce tokens from the diff range, but we can try to reconstruct original texts from the diffs and then check from which token in the list of tokens this text might start and end, and then reduce.

Reply:
Ah, because the diff is from the tokens to the result...